lookupapi/authentication.py

```diff
@@ -0,0 +1,104 @@
+"""
+OAuth2 authentication for Django REST Framework views.
+
+"""
+import datetime
+import logging
+
+from django.conf import settings
+from rest_framework.authentication import BaseAuthentication
+from requests_oauthlib import OAuth2Session
+from oauthlib.oauth2 import BackendApplicationClient, TokenExpiredError
+
+
+LOG = logging.getLogger()
+
+
+def _get_session():
+    """
+    Get a :py:class:`requests.Session` object which is authenticated with the API application's
+    OAuth2 client credentials.
+
+    """
+    client = BackendApplicationClient(client_id=settings.LOOKUP_API_OAUTH2_CLIENT_ID)
+    session = OAuth2Session(client=client)
+    session.fetch_token(
+        timeout=2, token_url=settings.LOOKUP_API_OAUTH2_TOKEN_URL,
+        client_id=settings.LOOKUP_API_OAUTH2_CLIENT_ID,
+        client_secret=settings.LOOKUP_API_OAUTH2_CLIENT_SECRET,
+        scope=settings.LOOKUP_API_OAUTH2_INTROSPECT_SCOPES)
+    return session
+
+
+def _request(*args, **kwargs):
+    """
+    A version of :py:func:`requests.request` which is authenticated with the OAuth2 token for the
+    API server's client credentials. If the token has timed out, it is requested again.
+
+    """
+    if getattr(_request, '__session', None) is None:
+        _request.__session = _get_session()
+    try:
+        return _request.__session.request(*args, **kwargs)
+    except TokenExpiredError:
+        _request.__session = _get_session()
+        return _request.__session.request(*args, **kwargs)
+
+
+def _utc_now():
+    """Return a UNIX-style timestamp representing "now" in UTC."""
+    return (datetime.datetime.utcnow() - datetime.datetime(1970, 1, 1)).total_seconds()
+
+
+class OAuth2TokenAuthentication(BaseAuthentication):
```

Comment:
Do we want to move this file to automationcommons? looks like its going to be used by more apps than just this

Reply:
I'd rather we let it bake a while in here to mature and then move it as a separate story. Some version of this will end up being used by our apps but it could do with some real-world testing first.